### DISCUSSION TITLE
Connecting to your mojo playground from VSCode or DataSpell

### DISCUSSION THREAD

**Discussion by strangemonad** [modular/modular#277] [Show and tell]
_This is less of a show and tell about something built in mojo and more about how to build something with mojo. Maybe a new topic type would be useful?_

## tl;dr;
- **Server URL** https://playground.modular.com/user/{USERNAME} e.g https://playground.modular.com/user/shawnmorel@gmail.com
- **Server Token** You can generate a token at https://playground.modular.com/hub/token

Note: I was able to get this working with Data Spell but not VSCode. I'm not as fluent with VSCode and in my experience with VSCode's Jupyter integration is that I tend to experience errors that don't get logged anywhere. It's probably something on my end. If someone can comment about VSCode working that would be great.

## Connecting to Jupyter Hub

Until there's a standalone toolchain we can run on our hardware, mojo code runs in [Modular's mojo playground](https://playground.modular.com/). Even once there's a standalone toolchain, hosted environments are convenient. 

If you're like me, without a bunch of Jupyter Lab extensions and configurations, I find the out of box Jupyter Lab editing experience painful. I also want to be able to use all my editor settings, auto-complete and GitHub Copilot. Thankfully, the mojo playground is hosted using [Jupyter Hub](https://jupyter.org/hub). Many IDEs and editors like VSCode, DataSpell, and PyCharm can connect to Jupyter Hub via its API.

- **Server URL** https://playground.modular.com/user/{USERNAME} e.g https://playground.modular.com/user/shawnmorel@gmail.com
- **Server Token** You can generate a token at https://playground.modular.com/hub/token
  - You can get here in the Jupyter Lab UI from `File > Hub Control Panel` https://playground.modular.com/hub/home 

Follow your IDE or editor workflow to connect to a remote Jupyter server
- **Data Spell** https://www.jetbrains.com/help/dataspell/configuring-jupyter-notebook.html Data Spell should even be able to automate fetching the token for you but you can always [generate a token manually](https://playground.modular.com/hub/token)
- **VSCode**  https://code.visualstudio.com/docs/datascience/jupyter-notebooks#_connect-to-a-remote-jupyter-server

https://github.com/modularml/mojo/assets/133905/98c9bd9c-c57c-4974-bfec-39f5708ec858

**Comment by mojodojodev** [comment]
[Here's how to do it from VS Code](https://mojodojo.dev/guides/general/mojo_playground_vscode.html)

**Reply by strangemonad** [reply]
thanks @mojodojodev. Loving the work you're doing. Any thoughts on how to make that content and discussions here easier to cross search? Happy to contribute this content back to mojo-dev too. My initial instinct is still to search these discussions.

I haven't looked at the Algolia API in a few years, maybe there's an easy way to have it search mojo-dev / mojo-dojo content and also include links to these discussions?

**Reply by mojodojodev** [reply]
Thanks appreciate that. I read everything that comes through issues and discussions in here, so can link if I have anything useful. For the other direction yes I should start adding links to these discussions and issues where appropriate good idea, not sure on Agolia I'll look into it.

**Reply by AboubakrN** [reply]
the link didnt work

**Reply by Moosems** [reply]
I signed into the playground using my email, what username and password should I use?

**Comment by BubbleDragonStop** [comment]
Thanks mate! I'm thinking of integrating it in my new project.
